docs/design_proposals/kaniko-logs-leakage.md

```diff
@@ -0,0 +1,74 @@
+# Skaffold logs improvements
+
+* Author(s): Prashant Arya (@prary)
+* Design Shepherd: 
+* Date: 7/05/2019
+* Status: 
+
+## Background
+
+Kaniko pod exiting before errors could be read by Skaffold. Do a normal skaffold run and put a invalid 
+base image or any other condition that would immediately(idea is to kill the kaniko pod before logger
+is attached ) fail to build a image and exit kaniko pod. In this condition user would not get acutual 
+reason of failure, he will just see following error
+
+build step: building [someBaseImageUrl]: kaniko build for [someBaseImageUrl]: waiting for pod to complete: pod already in terminal phase: Failed   
+
+
+Where as user should get actual error
+
+     Error while retrieving image from cache: could not parse reference
+     INFO[0002] Downloading base image gcr.io/invalidImageNameORUnReachableImage
+     error building image: could not parse reference
+
+
+This problem can also be exacted to a situation where user has put a such command which produce tons of output and before skaffold logger can pull all the logs kaniko pod  dies.
+
+## Design
+
+There should be a way where user can configure kaniko pod to stay alive for certain amount of time, so 
+that no logs are flushed before being fetched. There should be podGracePeriodSecond which ensures that
+kaniko pod stays alive for certain seconds after it finishes off building and pushing to registry. 
+
+Possible solution to counter above situation would be create a extra container in same kaniko pod
+which should continously monitor kaniko process. As soon as kaniko process terminate it should put 
+other container to sleep for specified mount of time. Extra container could be called a side car 
+which would not be very resource agnostic.
+
+### Kaniko Pod Definition changes
+
+```yaml
+Containers: []v1.Container{
+  {
+    Name:            constants.DefaultKanikoContainerName,
+    Image:           cfg.Image,
+    Args:            args,
+    ImagePullPolicy: v1.PullIfNotPresent,
+    Env:             []v1.EnvVar{},
+    VolumeMounts:    []v1.VolumeMount{},
+  },
+  {
+    Name:            "side-car",
+    Image:           constants.DefaultBusyboxImage,
+    ImagePullPolicy: v1.PullIfNotPresent,
+    Command: []string{"sh", "-c", "while [[ $(ps -ef | grep kaniko | wc -l) -gt 1 ]] ; do sleep 1; done; sleep " + cfg.PodGracePeriodSeconds},
```

Comment:
I see you specify this Container as a side-car container which adds a sleep until cgf.PodGracePeriodSeconds which is 2 seconds. 
Shd we instead use the preStopContainer hook to block container deletion ?
Shd we block it for number of seconds or something else is a question to discuss here.
```
lifecycle:
          preStop:
            exec:
              # SIGTERM triggers a quick exit; gracefully terminate instead
              command: ["some blocking command"]
```

Reply:
hi @tejal29 ,

IMHO I think preStopContainer are applicable only when api server sends kill pod command to a running pod, not when pod itself kills. Below is the link I have previously commented, correct me if i am missing something. We can have other solutions as well, I more of the option kaniko is the one where logs should be stored and streamed.

https://github.com/GoogleContainerTools/skaffold/issues/1978#issuecomment-489061535

Reply:
yes, i think we discussed about `kubectl describe` when we internally triaged #1978 . I just forgot about it. We can definitely rely on kubectl describe.  Should we do that when logs are not present?

Reply:
Correct me if I am wrong, you are pointing when kaniko is not producing any log in that case we should rely on kubectl describe, is that what you mean?

Reply:
yes. if we see an error or no lines when retrieving logs, we should run kubectl describe

Reply:
btw, Thanks @prary  for pointing out preStopContainers are only applicable when api server kills the pod.

Reply:
I think there is no need for kubectl describe we are already monitoring the status of pod in WaitForPodComplete function in https://github.com/GoogleContainerTools/skaffold/blob/master/pkg/skaffold/kubernetes/wait.go . Or we may just add a log in PodRunning switch case i.e 

```
                case v1.PodRunning:
                        logrus.Infof("Pod is running")
			return false, nil
		case v1.PodFailed:
			return false, fmt.Errorf("pod already in terminal phase: %s", pod.Status.Phase)
```

Reply:
Personally I like providing the output of kubectl describe upon failure because it could provide valuable information, and because the user won't be able to access it themselves since the pod will be cleaned up by skaffold.

Reply:
You mean 
```
case v1.PodFailed:
        **kubectl describe over here**
	return false, fmt.Errorf("pod already in terminal phase: %s", pod.Status.Phase)
```
Is that where you think kubectl describe should  be?

Reply:
@priyawadhwa @dgageot I think fetching events from kubernetes would be better approach than kubectl describe i.e
```
case v1.PodFailed:
   kubectl get events --namespace namespace-name --field-selector involvedObject.name=kaniko-pod-name
```
WDYT? This would fetch all event logs even when pod is failed due to some kubernetes constraints failure like image-builder(which we use for kaniko pod) service account not present or maybe attaching secret volume timeout due to some random reason.